### ISSUE TITLE
SDK 2.0 merge changes

### ISSUE THREAD

**Issue by hgray** [aws/aws-sdk-unity#60]
I'm finding some weird things happening with merging in the new SDK. I am using the Cognito Sync project to test.

- Deleted 'aws_cognito_cache.db' SQLite from local cache
- Deleted 'unity.Amazon.AWSUnitySyncDemo.plist' from preferences
- Created "characters" dataset in my authenticated ID in Cognito dashboard
- Added a single record with key "00200" and value "0.90;0.91;0.92;93" (key chosen to avoid conflict, same affect with 00000 key)
- Launched sync project and chose "Skip authentication"
- New identity created and stored in prefs, new dataset created in SQL
- Created two characters, clicked each to up the experience level
- Click "Save". DB now contains two characters:
     00000: 0.4590662;0.2431387;0.2483267;30
     00001: 0.8778638;0.3883582;0.2223839;57
- Repeated Save/Load shows no changes, I'm seeing two characters on screen
(So far, so good)

- Link with FB
- Screen updates to show a single character with an experience of 93 with a white color. This is the value that came from the authenticated characters. It didn't merge the two, it just is showing me the new one.
- Hit "Save". Now I have two characters, both with 93 experience level.
- Looking at the local db, it has duplicated the record so I now have two with a higher sync count, though I have only synced the "new" one once or so
- Every subsequent click of "Save" generates a new character with duplicate values to the original from my auth account (.90;.91;92;93)
- Clicking "Load" causes no problems
- Stopping app, restarting, logging in also causes a new character to be generated each time I do it
- At this point, my local DB has three datasets: my original unauth one ("characters"), my auth one ("characters"), and the other one I put into the auth account prior to syncing ("udder").
- On the record table, I'm seeing about what you'd expect: 2 records from my unauth user, and now 4 with my auth user.

I've gone through similar scenarios several times and I'm getting the same kind of results each time.
- I note that the merge is not triggered during the initial linking with authenticated account, nor with a call to save/sync or load, it is only triggered on a restart.
- Subsequent authenticated restarts will trigger the merge code again.
- Merging code is creating temp datasets with "unknown" in them, trying to merge, and then deleting, but the original data from my unauth identity remains. It seems that it's not actually renaming/merging the original unauth dataset.
- Exceptions are thrown during the merge flow as the "unknown" datasets are not found for syncing: "Dataset 'characters.unknown-1432070070.65555' not found."
"Recieved Exception of type ResourceNotFoundException:http://cognito-sync.amazonaws.com/doc/2014-06-30/, Message Dataset 'characters.unknown-1432070070.65555' not found., request id "

Let me know if you need more info.

![tables](https://cloud.githubusercontent.com/assets/2006948/7713897/cc49365e-fe48-11e4-9db1-0a9b7f3d685b.png)

**Comment by albertvaka** [comment]
Thanks for the report and the clear steps to reproduce the issue! The ResourceNotFoundException when trying to delete the merged dataset is a bug, and we are working on a fix. Once this is solved, you should be able to merge the datasets once and delete them, without having this problem where you end up merging them multiple times.

**Comment by hgray** [comment]
Sounds great, thanks!

**Comment by albertvaka** [comment]
The "infinite merge" problem happened when a dataset contained data created with a previous version of the sample (that used a slightly different format) and was imported into the new version of the sample. This should be fixed now. Note that we updated both the SDK [1] and the sample [2].

[1] http://aws.amazon.com/mobile/sdk/
[2] https://github.com/awslabs/aws-sdk-unity-samples

**Comment by hgray** [comment]
Thanks, I'll check it out.

**Comment by hgray** [comment]
I got this on one attempt. It did not trigger a sync failure:

An unhandled exception was thrown from the callback method <>m__5B.
UnityEngine.Debug:LogError(Object)
Amazon.Runtime.Internal.Util.UnityDebugLogger:Error(Exception, String, Object[]) (at Assets/AWSSDK/src/Core/Amazon.Runtime/Internal/Util/_unity/UnityDebugLogger.cs:45)
Amazon.Runtime.Internal.Util.Logger:Error(Exception, String, Object[]) (at Assets/AWSSDK/src/Core/Amazon.Runtime/Internal/Util/Logger.cs:155)
Amazon.Util.Internal.InternalSDKUtils:SafeExecute(Action) (at Assets/AWSSDK/src/Core/Amazon.Util/Internal/InternalSDKUtils.cs:157)
Amazon.Util.Internal.InternalSDKUtils:AsyncExecutor(Action, AsyncOptions) (at Assets/AWSSDK/src/Core/Amazon.Util/Internal/InternalSDKUtils.cs:124)
Amazon.CognitoSync.SyncManager.Internal.<PopulateListUpdatesAsync>c__AnonStorey41:<>m__54(AmazonServiceResult`2) (at Assets/AWSSDK/src/Services/CognitoSync/Custom/SyncManager/Storage/_unity/CognitoSyncStorage.cs:170)
Amazon.CognitoSync.<ListRecordsAsync>c__AnonStorey59:<>m__72(AmazonWebServiceRequest, AmazonWebServiceResponse, Exception, AsyncOptions) (at Assets/AWSSDK/src/Services/CognitoSync/Generated/_unity/AmazonCognitoSyncClient.cs:560)
Amazon.Runtime.Internal.UnityMainThreadDispatcher:ProcessRequests() (at Assets/AWSSDK/src/Core/Amazon.Runtime/Pipeline/_unity/UnityMainThreadDispatcher.cs:69)

**Comment by albertvaka** [comment]
Did this error happen following the same steps? I'm not able to reproduce it.

**Comment by hgray** [comment]
I get it every time I sign in with my authenticated account. I have cleared my local database and player prefs to ensure I wasn't getting any local oddness from past attempts. I also deleted all datasets from the account in the dashboard, so there shouldn't be any data in any accounts.

**Comment by albertvaka** [comment]
I'm still not able to reproduce it, I'm trying with the sample and sdk freshly download, and a brand new identity pool this time. I need some more information to reproduce this. What authentication provider are you using? Are you testing on a device or on the editor?

**Comment by hgray** [comment]
Clean version of sample and sdk. I'm still using my test identity pool which has an existing identity. I'm testing in Editor using Facebook authentication in the Cognito Sync Demo project.

And then I deleted all and redownloaded again. This time I emptied trash before pulling fresh. I don't know if it was holding on to something before, but this time it seems to work.

Sorry for the runaround. I'll keep you posted if anything pops up.
